drivers/meter/ade9000/ade9000.c

```diff
@@ -0,0 +1,528 @@
+					       * ADE9000_VOLTAGE_TR_FCN) / (uint64_t)ADE9000_RMS_FS_CODES));
+
+	ret = ade9000_read(dev, watt_reg, &temp);
+	if (ret)
+		return ret;
+
+	val_int = ADE9000_FS_VOLTAGE * (ADE9000_CURRENT_TR_FCN / 100)
+		  * (ADE9000_FS_VOLTAGE) * (ADE9000_VOLTAGE_TR_FCN / 10);
+	val_int2 = ADE9000_WATT_FS_CODES;
+	val_int = (uint64_t)((uint64_t)temp * (uint64_t)val_int);
+	val_int3 = (uint64_t)((uint64_t)val_int / (uint64_t)val_int2);
+	// Value in mW
+	dev->watt_val = (uint32_t)val_int3;
+
+	return 0;
+}
+
+/**
+ * @brief Set User Energy use model.
+ * @param dev - The device structure.
+ * @param model - The User Energy use model to be set.
+ * @param value - value specific to the user energy use model, it can be either
+ * 		1 for reading with reset, number of half line cycles or number
+ * 		of samples.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9000_set_egy_model(struct ade9000_dev *dev, enum ade9000_egy_model model,
+			  uint16_t value)
+{
+	int ret;
+	uint16_t egy_ld_accum, egy_tmr_mode, rd_rst_en, egy_pwr_en;
+
+	if (!dev)
+		return -ENODEV;
+
+	switch (model) {
+	case ADE9000_EGY_WITH_RESET:
+		/* For reading energy with reset the value should be 1 */
+		if (value != 1)
+			return -EINVAL;
+		egy_ld_accum = 0;
+		egy_tmr_mode = 0;
+		rd_rst_en = 1;
+		egy_pwr_en = 1;
+		break;
+	case ADE9000_EGY_HALF_LINE_CYCLES:
+		egy_ld_accum = 1;
+		egy_tmr_mode = 1;
+		rd_rst_en = 0;
+		egy_pwr_en = 1;
+		break;
+	case ADE9000_EGY_NR_SAMPLES:
+		egy_ld_accum = 1;
+		egy_tmr_mode = 0;
+		rd_rst_en = 0;
+		egy_pwr_en = 1;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = ade9000_write(dev, ADE9000_REG_RUN, 0);
+	if (ret)
+		return ret;
+
+	ret = ade9000_write(dev, ADE9000_REG_EP_CFG,
+			    no_os_field_prep(ADE9000_EGY_LD_ACCUM, egy_ld_accum) |
+			    no_os_field_prep(ADE9000_EGY_TMR_MODE, egy_tmr_mode) |
+			    no_os_field_prep(ADE9000_RD_RST_EN, rd_rst_en) |
+			    no_os_field_prep(ADE9000_EGY_PWR_EN, egy_pwr_en));
+	if (ret)
+		return ret;
+
+	ret = ade9000_write(dev, ADE9000_REG_EGY_TIME, value);
+	if (ret)
+		return ret;
+
+	return ade9000_write(dev, ADE9000_REG_RUN, 1);
+}
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade9000_init(struct ade9000_dev **device,
+		 struct ade9000_init_param init_param)
+{
+	int ret;
+	/* device structure */
+	struct ade9000_dev *dev;
+	/* chip id read value */
+	uint32_t chip_id;
+
+	dev = (struct ade9000_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	/* SPI Initialization*/
+	ret = no_os_spi_init(&dev->spi_desc, init_param.spi_init);
+	if (ret)
+		goto error_dev;
+
+	ret = ade9000_update_bits(dev, ADE9000_REG_CONFIG1, ADE9000_SWRST,
+				  no_os_field_prep(ADE9000_SWRST, 1));
+	if (ret)
+		goto error_spi;
+
+	no_os_mdelay(100);
```

Comment:
Is this 100ms value specified in the datasheet?

Reply:
In the datasheet the amount of time needed for power-on after reset is aprox. 47ms. The value is changed from 100ms to 50ms.